core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java

```diff
@@ -0,0 +1,167 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ThreadedActionListener;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class IndexShardOperationsLock {
+    private final ESLogger logger;
+    private final ThreadPool threadPool;
+
+    private static final int TOTAL_PERMITS = Integer.MAX_VALUE;
+    // fair semaphore to ensure that blockOperations() does not starve under thread contention
+    private final Semaphore semaphore = new Semaphore(TOTAL_PERMITS, true);
+    private final Object mutex = new Object(); // mutex used to protect the following two fields
+    private @Nullable List<ActionListener<Releasable>> delayedOperations; // operations that are delayed due to relocation hand-off
+    private boolean blockOperationsInProgress; // only allow one thread to try blocking operations at a time
+
+    public IndexShardOperationsLock(ESLogger logger, ThreadPool threadPool) {
+        this.logger = logger;
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Wait for in-flight operations to finish and executes onBlocked under the guarantee that no new operations are started. Queues
+     * operations that are occurring in the meanwhile and runs them once onBlocked has executed.
+     *
+     * @param timeout the maximum time to wait for the in-flight operations block
+     * @param timeUnit the time unit of the {@code timeout} argument
+     * @param onBlocked the action to run once the block has been acquired
+     * @throws IllegalStateException if operation is called while running
+     * @throws InterruptedException if calling thread is interrupted
+     * @throws TimeoutException if timed out waiting for in-flight operations to finish
+     */
+    public void blockOperations(long timeout, TimeUnit timeUnit, Runnable onBlocked) throws InterruptedException, TimeoutException {
```

Comment:
I this this can simplified even further : https://gist.github.com/bleskes/0bf520c969eaa9542b1deefb4a8e1d5a  (also note the test fixes, which are unrelated)